### DISCUSSION TITLE
Question about using references

### DISCUSSION THREAD

**Discussion by ntreeby** [Riverside-Software/pct#477] [Q&A] [answered]
I have a build.xml, and I am trying to use references so I can re-use my PCTConnection definitions.

However, I am seeing different behavior depending on how I set up my PCT taskdef.

For reference:
* Windows 10
* OpenEdge 11.7.6 (using %DLC%\pct\PCT.jar)
* java -version: openjdk version "11.0.13" 2021-10-19 LTS
OpenJDK Runtime Environment Microsoft-27990 (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM Microsoft-27990 (build 11.0.13+8-LTS, mixed mode)
* ant -version: Apache Ant(TM) version 1.9.8 compiled on December 25 2016 (using %DLC%\ant)
* DLC env var is set correctly

If I do this:
```
<project ...>
    <property environment="env"/>
    <taskdef resource="PCT.properties" classpath="${env.DLC}/pct/PCT.jar"/>
    <typedef resource="types.properties" classpath="${env.DLC}/pct/PCT.jar"/>
    
    <PCTConnection id="mydb" dbName="myDb" logicalName="myDb" hostName="somehost" dbPort="12345" readOnly="true"/>

    <target name="compile">
        <PCTCompile ...>
            <PCTConnection refid="mydb"/>
            ...
        </PCTCompile>
    </target>
</project>

```
Then, I get this error during the PCTCompile when it tries to reference the PCTConnection:
> ant compile
...
Class com.phenix.pct.PCTConnection (loaded via AntClassLoader[C:\Progress\OpenEdge11\pct\PCT.jar]) is not a subclass of com.phenix.pct.PCTConnection (loaded via AntClassLoader[C:\Progress\OpenEdge11\pct\PCT.jar])

BUILD FAILED
C:\Git\progress-monitor\build\build.xml:30: mydb doesn't denote a PCTConnection
...

However, if I change to this:
```
<project ... xmlns:pct="PCT:eu.rssw.pct">
    <property environment="env"/>
    <taskdef uri="PCT:eu.rssw.pct"
             resource="eu/rssw/pct/antlib.xml"
             classpath="${env.DLC}/pct/PCT.jar" />
    
    <pct:dbConnection id="mydb" dbName="myDb" logicalName="myDb" hostName="somehost" dbPort="12345" readOnly="true"/>

    <target name="compile">
        <pct:compile ...>
            <pct:dbConnection refid="mydb"/>
            ...
        </pct:compile>
    </target>
</project>

```
> ant compile
...
BUILD SUCCESSFUL
Total time: 2 seconds

Why is this? The Wiki suggests that the former approach will work, but it clearly isn't working for me.  Is there something I'm missing?

**Comment by gquerret** [comment, marked as answer]
Do you specify the classpath in the `<taskdef>` node ? If so, you should switch to `ant -lib /path/to/pct.jar`, or drop pct.jar in your $ANT_HOME/lib directory. You shouldn't see this problem anymore.

**Reply by ntreeby** [reply]
Yes, I do.  I thought I had C&P'd that into my original post, but I think some auto-formatting wiped it out.  Here is what I am using with my `<taskdef>` and `<typedef>` nodes:

```
    <taskdef resource="PCT.properties" classpath="${env.DLC}/pct/PCT.jar"/>
    <typedef resource="types.properties" classpath="${env.DLC}/pct/PCT.jar"/>

```
This works just fine for everything else, except when I use id="abc" / refid="abc" with PCT types.

**Reply by gquerret** [reply]
A few issues were already raised ( https://github.com/Riverside-Software/pct/issues?q=is%3Aissue+sort%3Aupdated-desc+AntClassLoader ), the classpath attribute shouldn't be set in your build file. I've removed the reference to classpath attribute in the installation guide.
